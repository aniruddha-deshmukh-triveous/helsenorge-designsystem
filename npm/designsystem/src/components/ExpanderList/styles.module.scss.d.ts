export type Styles = {
  'expander-list': string;
  'expander-list__item': string;
  'expander-list__item--line': string;
  'expander-list__item--outline': string;
  'expander-list-link': string;
  'expander-list-link__chevron': string;
  'expander-list-link__icon': string;
  'expander-list-link__main-content': string;
  'expander-list-link__main-content--expanded': string;
  'expander-list-link__main-content--padding': string;
  'expander-list-link--absolute': string;
  'expander-list-link--banana': string;
  'expander-list-link--blueberry': string;
  'expander-list-link--cherry': string;
  'expander-list-link--closed': string;
  'expander-list-link--fill': string;
  'expander-list-link--kiwi': string;
  'expander-list-link--large': string;
  'expander-list-link--neutral': string;
  'expander-list-link--plum': string;
  'expander-list-link--sticky': string;
  'expander-list-link--white': string;
};

export type ClassNames = keyof Styles;

declare const styles: Styles;

export default styles;
