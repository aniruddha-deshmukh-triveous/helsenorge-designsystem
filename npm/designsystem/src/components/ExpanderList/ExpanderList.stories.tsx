import React from 'react';
import { withKnobs, select, boolean } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import ExpanderList from './ExpanderList';
import LinkList from '../LinkList';
import Icon from '../Icons';
import Avatar from '../Icons/Avatar';
import PaperPlane from '../Icons/PaperPlane';
import AlarmClock from '../Icons/AlarmClock';
import { allPaletteNames } from '../../../.storybook/knobs';

const stories = storiesOf('ExpanderList', module);
stories.addDecorator(withKnobs);

stories.add('Default', () => (
  <div
    style={{
      width: '40rem',
    }}
  >
    <ExpanderList
      isOpen={boolean('Is open', false)}
      accordion={boolean('Accordion', false)}
      childPadding={boolean('Child padding', true)}
      color={select('Color', allPaletteNames, 'blueberry')}
      topBorder={boolean('Top border', true)}
      bottomBorder={boolean('Bottom border', true)}
      large={boolean('Large', false)}
    >
      <ExpanderList.Expander title="Kognitiv terapi">
        Kognitiv terapi er en form for psykoterapi som retter seg mot problemløsning og innsikt i sammenhengen mellom tenkning, handlinger
        og følelser. Et viktig mål er å bryte selvforsterkende onde sirkler som opprettholder psykiske helseproblemer.
      </ExpanderList.Expander>
      <ExpanderList.Expander title="Hypokondri">
        Hypokondri er en sykdom der folk føler at de har en sykdom som de i realiteten ikke har. Statens helsetilsyn sier blant annet
        følgende om sykdommen: «Det vesentlige kjennetegnet er vedvarende opptatthet av muligheten for å ha en eller flere alvorlige og
        fremadskridende somatiske lidelser».
      </ExpanderList.Expander>
      <ExpanderList.Expander title="En hjerneskade er en skade opstået i hjernen, med vedvarende funktionsnedsættelse til følge (Ekstra lang tekst for wrapping)">
        De hyppigste årsager til hjerneskader er hjerneblødninger, blodpropper i hjernen, trafik- eller drukneulykker, svulster eller
        hjertestop med efterfølgende iltmangel til hjernen; men kan også skyldes en hjernebetændelse på grund af herpes eller anden virus.
      </ExpanderList.Expander>
    </ExpanderList>
  </div>
));

stories.add('With icon', () => (
  <div
    style={{
      width: '40rem',
    }}
  >
    <ExpanderList isOpen={boolean('Is open', false)} color="cherry">
      <ExpanderList.Expander icon={<Icon svgIcon={Avatar} />} title="Kognitiv terapi">
        Kognitiv terapi er en form for psykoterapi som retter seg mot problemløsning og innsikt i sammenhengen mellom tenkning, handlinger
        og følelser. Et viktig mål er å bryte selvforsterkende onde sirkler som opprettholder psykiske helseproblemer.
      </ExpanderList.Expander>
      <ExpanderList.Expander icon={<Icon svgIcon={PaperPlane} />} title="Hypokondri">
        Hypokondri er en sykdom der folk føler at de har en sykdom som de i realiteten ikke har. Statens helsetilsyn sier blant annet
        følgende om sykdommen: «Det vesentlige kjennetegnet er vedvarende opptatthet av muligheten for å ha en eller flere alvorlige og
        fremadskridende somatiske lidelser».
      </ExpanderList.Expander>
      <ExpanderList.Expander icon={<Icon svgIcon={AlarmClock} />} title="Hjerneskade">
        De hyppigste årsager til hjerneskader er hjerneblødninger, blodpropper i hjernen, trafik- eller drukneulykker, svulster eller
        hjertestop med efterfølgende iltmangel til hjernen; men kan også skyldes en hjernebetændelse på grund af herpes eller anden virus.
      </ExpanderList.Expander>
    </ExpanderList>
  </div>
));

stories.add('With LinkList', () => (
  <div
    style={{
      width: '40rem',
    }}
  >
    <ExpanderList isOpen={boolean('Is open', false)} color="cherry">
      <ExpanderList.Expander title="Kognitiv terapi">
        <LinkList color="cherry">
          <LinkList.Link href="/kognitivterapi1">Første lenke</LinkList.Link>
          <LinkList.Link href="/kognitivterapi2">Andre lenke</LinkList.Link>
          <LinkList.Link href="/kognitivterapi3">Tredje lenke</LinkList.Link>
        </LinkList>
      </ExpanderList.Expander>
    </ExpanderList>
  </div>
));
stories.add('As accordion', () => (
  <div
    style={{
      width: '40rem',
    }}
  >
    <p>
      Nær 1,6 million mennesker i Norge tilhører grupper med økt risiko for komplikasjoner av influensa. Det er anslått at det i
      gjennomsnitt dør 900 personer i Norge årlig som følge av sykdommen. Influensavaksine kan beskytte mange av disse.
    </p>
    <p>
      Influensa kan blant annet føre til lungebetennelse og forverring av kroniske sykdommer. Personer med hjerte-/karlidelser er i
      influensasesongen mer utsatt for hjerteinfarkt, hjerneslag og død. Ved alvorlige komplikasjoner av influensa kan sykehusinnleggelse
      være nødvendig. Noen får varig svekket helse etter alvorlig influensasykdom. Vaksinasjon kan bidra til å beskytte mot slike hendelser.
    </p>
    <p>
      Influensa i svangerskapet gir noe økt risiko for dødfødsel. Gravide er også mer utsatt for følgesykdommer som lungebetennelse enn
      andre friske kvinner. Influensa hos spebarn kan være alvorlig. Vaksinasjon av mor under graviditeten beskytter mor under svangerskapet
      og gir også barnet beskyttelse mot influensa den første tiden etter fødselen.
    </p>
    <ExpanderList accordion={boolean('Accordion', true)}>
      <ExpanderList.Expander title="Influensa">
        <p>
          Nær 1,6 million mennesker i Norge tilhører grupper med økt risiko for komplikasjoner av influensa. Det er anslått at det i
          gjennomsnitt dør 900 personer i Norge årlig som følge av sykdommen. Influensavaksine kan beskytte mange av disse.
        </p>
        <p>
          Influensa kan blant annet føre til lungebetennelse og forverring av kroniske sykdommer. Personer med hjerte-/karlidelser er i
          influensasesongen mer utsatt for hjerteinfarkt, hjerneslag og død. Ved alvorlige komplikasjoner av influensa kan
          sykehusinnleggelse være nødvendig. Noen får varig svekket helse etter alvorlig influensasykdom. Vaksinasjon kan bidra til å
          beskytte mot slike hendelser.
        </p>
        <p>
          Influensa i svangerskapet gir noe økt risiko for dødfødsel. Gravide er også mer utsatt for følgesykdommer som lungebetennelse enn
          andre friske kvinner. Influensa hos spebarn kan være alvorlig. Vaksinasjon av mor under graviditeten beskytter mor under
          svangerskapet og gir også barnet beskyttelse mot influensa den første tiden etter fødselen.
        </p>
      </ExpanderList.Expander>
      <ExpanderList.Expander title="Disse bør ta influensavaksinen">
        <p>
          Nær 1,6 million mennesker i Norge tilhører grupper med økt risiko for komplikasjoner av influensa. Det er anslått at det i
          gjennomsnitt dør 900 personer i Norge årlig som følge av sykdommen. Influensavaksine kan beskytte mange av disse.
        </p>
        <p>
          Influensa kan blant annet føre til lungebetennelse og forverring av kroniske sykdommer. Personer med hjerte-/karlidelser er i
          influensasesongen mer utsatt for hjerteinfarkt, hjerneslag og død. Ved alvorlige komplikasjoner av influensa kan
          sykehusinnleggelse være nødvendig. Noen får varig svekket helse etter alvorlig influensasykdom. Vaksinasjon kan bidra til å
          beskytte mot slike hendelser.
        </p>
        <p>
          Influensa i svangerskapet gir noe økt risiko for dødfødsel. Gravide er også mer utsatt for følgesykdommer som lungebetennelse enn
          andre friske kvinner. Influensa hos spebarn kan være alvorlig. Vaksinasjon av mor under graviditeten beskytter mor under
          svangerskapet og gir også barnet beskyttelse mot influensa den første tiden etter fødselen.
        </p>
      </ExpanderList.Expander>
      <ExpanderList.Expander title="Beskytter mot flere influensavirus, men må tas årlig">
        <p>
          Nær 1,6 million mennesker i Norge tilhører grupper med økt risiko for komplikasjoner av influensa. Det er anslått at det i
          gjennomsnitt dør 900 personer i Norge årlig som følge av sykdommen. Influensavaksine kan beskytte mange av disse.
        </p>
        <p>
          Influensa kan blant annet føre til lungebetennelse og forverring av kroniske sykdommer. Personer med hjerte-/karlidelser er i
          influensasesongen mer utsatt for hjerteinfarkt, hjerneslag og død. Ved alvorlige komplikasjoner av influensa kan
          sykehusinnleggelse være nødvendig. Noen får varig svekket helse etter alvorlig influensasykdom. Vaksinasjon kan bidra til å
          beskytte mot slike hendelser.
        </p>
        <p>
          Influensa i svangerskapet gir noe økt risiko for dødfødsel. Gravide er også mer utsatt for følgesykdommer som lungebetennelse enn
          andre friske kvinner. Influensa hos spebarn kan være alvorlig. Vaksinasjon av mor under graviditeten beskytter mor under
          svangerskapet og gir også barnet beskyttelse mot influensa den første tiden etter fødselen.
        </p>
      </ExpanderList.Expander>
      <ExpanderList.Expander title="Bivirkninger av influensavaksinen">
        <p>
          Nær 1,6 million mennesker i Norge tilhører grupper med økt risiko for komplikasjoner av influensa. Det er anslått at det i
          gjennomsnitt dør 900 personer i Norge årlig som følge av sykdommen. Influensavaksine kan beskytte mange av disse.
        </p>
        <p>
          Influensa kan blant annet føre til lungebetennelse og forverring av kroniske sykdommer. Personer med hjerte-/karlidelser er i
          influensasesongen mer utsatt for hjerteinfarkt, hjerneslag og død. Ved alvorlige komplikasjoner av influensa kan
          sykehusinnleggelse være nødvendig. Noen får varig svekket helse etter alvorlig influensasykdom. Vaksinasjon kan bidra til å
          beskytte mot slike hendelser.
        </p>
        <p>
          Influensa i svangerskapet gir noe økt risiko for dødfødsel. Gravide er også mer utsatt for følgesykdommer som lungebetennelse enn
          andre friske kvinner. Influensa hos spebarn kan være alvorlig. Vaksinasjon av mor under graviditeten beskytter mor under
          svangerskapet og gir også barnet beskyttelse mot influensa den første tiden etter fødselen.
        </p>
      </ExpanderList.Expander>
    </ExpanderList>
  </div>
));

stories.add('With callback', () => (
  <div
    style={{
      width: '40rem',
    }}
  >
    <ExpanderList
      isOpen={boolean('Is open', false)}
      accordion={boolean('Accordion', false)}
      childPadding={boolean('Child padding', true)}
      color={select('Color', allPaletteNames, 'blueberry')}
      topBorder={boolean('Top border', true)}
      bottomBorder={boolean('Bottom border', true)}
      large={boolean('Large', false)}
    >
      <ExpanderList.Expander title="Kognitiv terapi" onExpand={isExpanded => console.log(isExpanded)}>
        <p>Sjekk nettleserkonsollen</p>
      </ExpanderList.Expander>
    </ExpanderList>
  </div>
));
