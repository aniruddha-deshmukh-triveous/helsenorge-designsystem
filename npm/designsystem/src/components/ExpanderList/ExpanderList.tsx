import React, { useEffect, useState } from 'react';

import { PaletteNames } from '../../theme/palette';
import Icon, { IconSize } from '../Icons';
import ChevronUp from '../Icons/ChevronUp';
import ChevronDown from '../Icons/ChevronDown';
import { useHover } from '../../hooks/useHover';
import { usePrevious } from '../../hooks/usePrevious';
import { Breakpoint, useBreakpoint } from '../../hooks/useBreakpoint';
import { isElementInViewport } from '../../utils/viewport';

import classNames from 'classnames';

import expanderListStyles from './styles.module.scss';
import { AnalyticsId } from '../../constants';
import { useUuid } from '../../hooks/useUuid';

export type ExpanderListColors = PaletteNames;
export type ExpanderType = React.ForwardRefExoticComponent<ExpanderProps & React.RefAttributes<HTMLLIElement>>;

export interface ExpanderListCompound extends React.ForwardRefExoticComponent<ExpanderListProps & React.RefAttributes<HTMLUListElement>> {
  Expander: ExpanderType;
}

interface ExpanderListProps {
  /** Toggles accordion functionality for the expanders. */
  accordion?: boolean;
  /** Toggles the bottom border of the last child element. */
  bottomBorder?: boolean;
  /** Items in the ExpanderList */
  children: React.ReactNode;
  /** Toggles padding of child elements */
  childPadding?: boolean;
  /** Adds custom classes to the element. */
  className?: string;
  /** Changes the link list background color on hover. */
  color?: ExpanderListColors;
  /** Changes the font size. */
  large?: boolean;
  /** Opens the first item in the list. */
  isOpen?: boolean;
  /** Toggles the top border of the first child element. */
  topBorder?: boolean;
  /** Sets the data-testid attribute. */
  testId?: string;
}

type Modify<T, R> = Omit<T, keyof R> & R;

type ExpanderProps = Modify<
  React.HTMLAttributes<HTMLButtonElement>,
  {
    id?: string;
    title: JSX.Element | string;
    children: React.ReactNode;
    className?: string;
    color?: ExpanderListColors;
    icon?: React.ReactElement;
    padding?: boolean;
    expanded?: boolean;
    large?: boolean;
    testId?: string;
    handleExpanderClick?: (event: React.MouseEvent<HTMLElement, MouseEvent>) => void;
    /** Called when expander is open/closed. */
    onExpand?: (isExpanded: boolean) => void;
  }
>;

// TODO: See what can be done with regards to double reffing.
const Expander: ExpanderType = React.forwardRef((props: ExpanderProps, ref: React.Ref<HTMLLIElement>) => {
  const {
    id,
    children,
    padding = true,
    color = 'neutral',
    className = '',
    icon,
    large = false,
    title,
    expanded = false,
    testId,
    handleExpanderClick,
    onExpand,
  } = props;
  const [isExpanded, setIsExpanded] = useState<boolean>(expanded);
  const previousIsExpanded = usePrevious(isExpanded);
  const { hoverRef, isHovered } = useHover<HTMLButtonElement>();
  const breakpoint = useBreakpoint();
  const expanderClasses = classNames(expanderListStyles['expander-list-link'], expanderListStyles[`expander-list-link--${color}`], {
    [expanderListStyles['expander-list-link--closed']]: !isExpanded,
    [expanderListStyles['expander-list-link--large']]: large,
  });
  const mainContentClasses = classNames(
    expanderListStyles['expander-list-link__main-content'],
    isExpanded && expanderListStyles['expander-list-link__main-content--expanded'],
    padding ? expanderListStyles['expander-list-link__main-content--padding'] : ''
  );

  useEffect(() => {
    if (expanded !== isExpanded) {
      setIsExpanded(expanded);
    }
  }, [expanded]);

  useEffect(() => {
    if (onExpand && isExpanded !== !!previousIsExpanded) {
      onExpand(isExpanded);
    }
  }, [isExpanded, onExpand]);

  return (
    <li className={className} ref={ref}>
      <button
        id={id}
        onClick={handleExpanderClick}
        data-testid={testId}
        data-analyticsid={AnalyticsId.ExpanderListExpander}
        className={expanderClasses}
        ref={hoverRef}
        aria-expanded={isExpanded}
      >
        {icon && (
          <span className={expanderListStyles['expander-list-link__icon']}>
            {React.cloneElement(icon, {
              size: breakpoint === Breakpoint.xs ? IconSize.XSmall : IconSize.Small,
              isHovered,
            })}
          </span>
        )}
        <span className={expanderListStyles['expander-list-link__title']}>{title}</span>
        <span className={expanderListStyles['expander-list-link__chevron']}>
          <Icon size={IconSize.XSmall} svgIcon={isExpanded ? ChevronUp : ChevronDown} isHovered={isHovered} />
        </span>
      </button>
      <div className={mainContentClasses}>{children}</div>
    </li>
  );
});

export const ExpanderList = React.forwardRef((props: ExpanderListProps, ref: React.Ref<HTMLUListElement>) => {
  const {
    children,
    childPadding = true,
    large,
    isOpen = false,
    color,
    className = '',
    accordion = false,
    topBorder = true,
    bottomBorder = true,
    testId,
  } = props;
  const [activeExpander, setActiveExpander] = useState({});
  const [latestExpander, setLatestExpander] = useState<HTMLElement>();
  const uuid = useUuid();
  const childCount = React.Children.count(children);
  const expanderListClasses = classNames(expanderListStyles['expander-list'], className);

  function handleExpanderClick(event: React.MouseEvent<HTMLElement, MouseEvent>, id: string): void {
    setActiveExpander(prevState => (accordion ? { [id]: !prevState[id] } : { ...prevState, [id]: !prevState[id] }));
    setLatestExpander(event.currentTarget);
  }

  /** Returns the class modifier top when we want to show the top border and no-bottom when we don't want to show the bottom border */
  const getExpanderItemClass = (index: number) => {
    return classNames(expanderListStyles['expander-list__item'], {
      [expanderListStyles['expander-list__item--top']]: index === 0 && topBorder,
      [expanderListStyles['expander-list__item--no-bottom']]: index === childCount - 1 && !bottomBorder,
    });
  };

  useEffect(() => {
    if (accordion && latestExpander && !isElementInViewport(latestExpander)) {
      latestExpander.scrollIntoView();
    }
  }, [accordion, latestExpander]);

  useEffect(() => {
    if (isOpen) {
      const id = `${uuid}-0`;
      setActiveExpander(prevState => (accordion ? { [id]: !prevState[id] } : { ...prevState, [id]: !prevState[id] }));
    }
  }, [isOpen]);

  return (
    <ul className={expanderListClasses} ref={ref} data-testid={testId} data-analyticsid={AnalyticsId.ExpanderList}>
      {React.Children.map(children, (child: React.ReactNode, index: number) => {
        if ((child as React.ReactElement<ExpanderProps>).type === Expander) {
          const expanded = activeExpander[`${uuid}-${index}`];

          const expanderItemClass = getExpanderItemClass(index);

          return React.cloneElement(child as React.ReactElement<ExpanderProps>, {
            id: `${uuid}-${index}`,
            key: index,
            expanded,
            padding: childPadding,
            color,
            large,
            'aria-expanded': expanded,
            className: expanderItemClass,
            handleExpanderClick: (event: React.MouseEvent<HTMLElement>) => handleExpanderClick(event, `${uuid}-${index}`),
          });
        }
        return child;
      })}
    </ul>
  );
}) as ExpanderListCompound;

ExpanderList.Expander = Expander;
Expander.displayName = 'ExpanderList.Expander';

export default ExpanderList;
