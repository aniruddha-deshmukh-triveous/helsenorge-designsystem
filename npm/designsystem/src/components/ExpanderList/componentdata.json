{
  "props": {
    "accordion": {
      "defaultValue": null,
      "description": "Toggles accordion functionality for the expanders.",
      "name": "accordion",
      "parent": {
        "fileName": "src/components/ExpanderList/ExpanderList.tsx",
        "name": "ExpanderListProps"
      },
      "declarations": [
        {
          "fileName": "src/components/ExpanderList/ExpanderList.tsx",
          "name": "ExpanderListProps"
        }
      ],
      "required": false,
      "type": {
        "name": "boolean"
      }
    },
    "bottomBorder": {
      "defaultValue": null,
      "description": "Toggles the bottom border of the last child element.",
      "name": "bottomBorder",
      "parent": {
        "fileName": "src/components/ExpanderList/ExpanderList.tsx",
        "name": "ExpanderListProps"
      },
      "declarations": [
        {
          "fileName": "src/components/ExpanderList/ExpanderList.tsx",
          "name": "ExpanderListProps"
        }
      ],
      "required": false,
      "type": {
        "name": "boolean"
      }
    },
    "children": {
      "defaultValue": null,
      "description": "Items in the ExpanderList",
      "name": "children",
      "parent": {
        "fileName": "src/components/ExpanderList/ExpanderList.tsx",
        "name": "ExpanderListProps"
      },
      "declarations": [
        {
          "fileName": "src/components/ExpanderList/ExpanderList.tsx",
          "name": "ExpanderListProps"
        }
      ],
      "required": true,
      "type": {
        "name": "ReactNode"
      }
    },
    "childPadding": {
      "defaultValue": null,
      "description": "Toggles padding of child elements",
      "name": "childPadding",
      "parent": {
        "fileName": "src/components/ExpanderList/ExpanderList.tsx",
        "name": "ExpanderListProps"
      },
      "declarations": [
        {
          "fileName": "src/components/ExpanderList/ExpanderList.tsx",
          "name": "ExpanderListProps"
        }
      ],
      "required": false,
      "type": {
        "name": "boolean"
      }
    },
    "className": {
      "defaultValue": null,
      "description": "Adds custom classes to the element.",
      "name": "className",
      "parent": {
        "fileName": "src/components/ExpanderList/ExpanderList.tsx",
        "name": "ExpanderListProps"
      },
      "declarations": [
        {
          "fileName": "src/components/ExpanderList/ExpanderList.tsx",
          "name": "ExpanderListProps"
        }
      ],
      "required": false,
      "type": {
        "name": "string"
      }
    },
    "color": {
      "defaultValue": null,
      "description": "Changes the link list background color on hover.",
      "name": "color",
      "parent": {
        "fileName": "src/components/ExpanderList/ExpanderList.tsx",
        "name": "ExpanderListProps"
      },
      "declarations": [
        {
          "fileName": "src/components/ExpanderList/ExpanderList.tsx",
          "name": "ExpanderListProps"
        }
      ],
      "required": false,
      "type": {
        "name": "enum",
        "raw": "\"banana\" | \"blueberry\" | \"cherry\" | \"kiwi\" | \"neutral\" | \"plum\" | \"black\" | \"white\" | undefined",
        "value": [
          {
            "value": "undefined"
          },
          {
            "value": "\"banana\""
          },
          {
            "value": "\"blueberry\""
          },
          {
            "value": "\"cherry\""
          },
          {
            "value": "\"kiwi\""
          },
          {
            "value": "\"neutral\""
          },
          {
            "value": "\"plum\""
          },
          {
            "value": "\"black\""
          },
          {
            "value": "\"white\""
          }
        ]
      }
    },
    "large": {
      "defaultValue": null,
      "description": "Changes the font size.",
      "name": "large",
      "parent": {
        "fileName": "src/components/ExpanderList/ExpanderList.tsx",
        "name": "ExpanderListProps"
      },
      "declarations": [
        {
          "fileName": "src/components/ExpanderList/ExpanderList.tsx",
          "name": "ExpanderListProps"
        }
      ],
      "required": false,
      "type": {
        "name": "boolean"
      }
    },
    "isOpen": {
      "defaultValue": null,
      "description": "Opens the first item in the list.",
      "name": "isOpen",
      "parent": {
        "fileName": "src/components/ExpanderList/ExpanderList.tsx",
        "name": "ExpanderListProps"
      },
      "declarations": [
        {
          "fileName": "src/components/ExpanderList/ExpanderList.tsx",
          "name": "ExpanderListProps"
        }
      ],
      "required": false,
      "type": {
        "name": "boolean"
      }
    },
    "topBorder": {
      "defaultValue": null,
      "description": "Toggles the top border of the first child element.",
      "name": "topBorder",
      "parent": {
        "fileName": "src/components/ExpanderList/ExpanderList.tsx",
        "name": "ExpanderListProps"
      },
      "declarations": [
        {
          "fileName": "src/components/ExpanderList/ExpanderList.tsx",
          "name": "ExpanderListProps"
        }
      ],
      "required": false,
      "type": {
        "name": "boolean"
      }
    }
  }
}
