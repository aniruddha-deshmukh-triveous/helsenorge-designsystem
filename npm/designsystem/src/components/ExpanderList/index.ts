import ExpanderList from './ExpanderList';
export * from './ExpanderList';
export default ExpanderList;
