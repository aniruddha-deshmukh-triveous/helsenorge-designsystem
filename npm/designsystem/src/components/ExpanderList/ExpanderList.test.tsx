import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import ExpanderList from './ExpanderList';
import { theme } from '../../theme';
import * as ViewportUtils from '../../utils/viewport';

describe('Gitt ExpanderList blir rendret', (): void => {
  describe('Når det er tre expandere og accordion er false', (): void => {
    it('Sjekk at ExpanderList sitt snapshot matcher', (): void => {
      const { container } = render(
        <div
          style={{
            width: '40rem',
          }}
        >
          <ExpanderList isOpen={false} accordion={false} childPadding={true} color={'blueberry'}>
            <ExpanderList.Expander title="Title 1">Text 1</ExpanderList.Expander>
            <ExpanderList.Expander title="Title 2">Text 2</ExpanderList.Expander>
            <ExpanderList.Expander title="Title 3">Text 2</ExpanderList.Expander>
          </ExpanderList>
        </div>
      );
      expect(container).toMatchSnapshot();
    });
  });

  describe('Når man klikker på en expander', (): void => {
    it('Sjekk Expander bakgrunnsfarge når musepeker trykker på den', (): void => {
      render(
        <div
          style={{
            width: '40rem',
          }}
        >
          <ExpanderList isOpen={false} accordion={false} childPadding={true} color={'blueberry'}>
            <ExpanderList.Expander testId="test01" title="Title 1">
              Text 1
            </ExpanderList.Expander>
            <ExpanderList.Expander title="Title 2">Text 2</ExpanderList.Expander>
            <ExpanderList.Expander title="Title 3">Text 2</ExpanderList.Expander>
          </ExpanderList>
        </div>
      );

      const testExpander = screen.getByTestId('test01');
      fireEvent.click(testExpander);
      expect(testExpander.className).toBe('expander-list-link expander-list-link--blueberry');
    });
  });

  describe('Når accordion er true', (): void => {
    test('Så vises bare en tekst om gangen', (): void => {
      render(
        <ExpanderList accordion={true}>
          <ExpanderList.Expander title="Title 1">Text 1</ExpanderList.Expander>
          <ExpanderList.Expander title="Title 2">Text 2</ExpanderList.Expander>
          <ExpanderList.Expander title="Title 3">Text 2</ExpanderList.Expander>
        </ExpanderList>
      );

      const title1 = screen.getByRole('button', { name: 'Title 1' });
      userEvent.click(title1);
      const text1 = screen.getByText('Text 1');
      expect(text1).toBeVisible();
      const title2 = screen.getByRole('button', { name: 'Title 2' });
      userEvent.click(title2);
      expect(screen.queryByText('Text 1')).not.toBeInTheDocument();
      const text2 = screen.getByText('Text 2');
      expect(text2).toBeVisible();
    });
  });

  describe('Når accordion er true og expanderen er utenfor viewport etter at man klikker på den', (): void => {
    test('Så scroller nettleseren slik at expanderen blir synlig', (): void => {
      const mockScrollIntoView = jest.fn();
      window.HTMLElement.prototype.scrollIntoView = mockScrollIntoView;
      const mockIsElementInViewport = jest.spyOn(ViewportUtils, 'isElementInViewport');

      mockIsElementInViewport.mockReturnValue(false);
      render(
        <ExpanderList accordion={true}>
          <ExpanderList.Expander title="Title 1">Text 1</ExpanderList.Expander>
          <ExpanderList.Expander title="Title 2">Text 2</ExpanderList.Expander>
          <ExpanderList.Expander title="Title 3">Text 2</ExpanderList.Expander>
        </ExpanderList>
      );

      const title1 = screen.getByRole('button', { name: 'Title 3' });
      userEvent.click(title1);
      expect(mockScrollIntoView).toHaveBeenCalledTimes(1);
      expect(mockScrollIntoView).toHaveBeenCalledWith();
    });
  });
});
