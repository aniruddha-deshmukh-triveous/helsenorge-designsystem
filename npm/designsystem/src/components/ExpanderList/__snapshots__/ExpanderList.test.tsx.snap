// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gitt ExpanderList blir rendret Når det er to ExpanderList i samme dokument Så har expanderne forskjellige IDer 1`] = `
<div>
  <ul
    class="expander-list"
    data-analyticsid="expander-list"
  >
    <li
      class="expander-list__item expander-list__item--top"
    >
      <button
        aria-expanded="false"
        class="expander-list-link expander-list-link--neutral expander-list-link--closed"
        data-analyticsid="expander-list-expander"
        id="testid-35-0"
        type="button"
      >
        <span
          class="expander-list-link__title"
        >
          ExpanderList 1
        </span>
        <span
          class="expander-list-link__chevron"
        >
          <svg
            aria-hidden="true"
            class="hnds-style-icon "
            data-analyticsid="icon"
            fill="black"
            height="38"
            role="img"
            style="min-width: 38px; min-height: 38px;"
            viewBox="0 0 48 48"
            width="38"
          >
            <path
              d="M32.577 17.885l1.793 1.779-10.371 10.451L13.63 19.664l1.793-1.779 8.576 8.644z"
            />
          </svg>
        </span>
      </button>
      <div
        class="expander-list-link__main-content expander-list-link__main-content--padding"
      >
        Text 1
      </div>
    </li>
  </ul>
  <ul
    class="expander-list"
    data-analyticsid="expander-list"
  >
    <li
      class="expander-list__item expander-list__item--top"
    >
      <button
        aria-expanded="false"
        class="expander-list-link expander-list-link--neutral expander-list-link--closed"
        data-analyticsid="expander-list-expander"
        id="testid-36-0"
        type="button"
      >
        <span
          class="expander-list-link__title"
        >
          ExpanderList 2
        </span>
        <span
          class="expander-list-link__chevron"
        >
          <svg
            aria-hidden="true"
            class="hnds-style-icon "
            data-analyticsid="icon"
            fill="black"
            height="38"
            role="img"
            style="min-width: 38px; min-height: 38px;"
            viewBox="0 0 48 48"
            width="38"
          >
            <path
              d="M32.577 17.885l1.793 1.779-10.371 10.451L13.63 19.664l1.793-1.779 8.576 8.644z"
            />
          </svg>
        </span>
      </button>
      <div
        class="expander-list-link__main-content expander-list-link__main-content--padding"
      >
        Text 2
      </div>
    </li>
  </ul>
</div>
`;

exports[`Gitt ExpanderList blir rendret Når det er tre expandere og accordion er false Sjekk at ExpanderList sitt snapshot matcher 1`] = `
<div>
  <ul
    class="expander-list"
    data-analyticsid="expander-list"
  >
    <li
      class="expander-list__item expander-list__item--top"
    >
      <button
        aria-expanded="false"
        class="expander-list-link expander-list-link--blueberry expander-list-link--closed"
        data-analyticsid="expander-list-expander"
        id="testid-1-0"
        type="button"
      >
        <span
          class="expander-list-link__title"
        >
          Title 1
        </span>
        <span
          class="expander-list-link__chevron"
        >
          <svg
            aria-hidden="true"
            class="hnds-style-icon "
            data-analyticsid="icon"
            fill="black"
            height="38"
            role="img"
            style="min-width: 38px; min-height: 38px;"
            viewBox="0 0 48 48"
            width="38"
          >
            <path
              d="M32.577 17.885l1.793 1.779-10.371 10.451L13.63 19.664l1.793-1.779 8.576 8.644z"
            />
          </svg>
        </span>
      </button>
      <div
        class="expander-list-link__main-content expander-list-link__main-content--padding"
      >
        Text 1
      </div>
    </li>
    <li
      class="expander-list__item"
    >
      <button
        aria-expanded="false"
        class="expander-list-link expander-list-link--blueberry expander-list-link--closed"
        data-analyticsid="expander-list-expander"
        id="testid-1-1"
        type="button"
      >
        <span
          class="expander-list-link__title"
        >
          Title 2
        </span>
        <span
          class="expander-list-link__chevron"
        >
          <svg
            aria-hidden="true"
            class="hnds-style-icon "
            data-analyticsid="icon"
            fill="black"
            height="38"
            role="img"
            style="min-width: 38px; min-height: 38px;"
            viewBox="0 0 48 48"
            width="38"
          >
            <path
              d="M32.577 17.885l1.793 1.779-10.371 10.451L13.63 19.664l1.793-1.779 8.576 8.644z"
            />
          </svg>
        </span>
      </button>
      <div
        class="expander-list-link__main-content expander-list-link__main-content--padding"
      >
        Text 2
      </div>
    </li>
    <li
      class="expander-list__item"
    >
      <button
        aria-expanded="false"
        class="expander-list-link expander-list-link--blueberry expander-list-link--closed"
        data-analyticsid="expander-list-expander"
        id="testid-1-2"
        type="button"
      >
        <span
          class="expander-list-link__title"
        >
          Title 3
        </span>
        <span
          class="expander-list-link__chevron"
        >
          <svg
            aria-hidden="true"
            class="hnds-style-icon "
            data-analyticsid="icon"
            fill="black"
            height="38"
            role="img"
            style="min-width: 38px; min-height: 38px;"
            viewBox="0 0 48 48"
            width="38"
          >
            <path
              d="M32.577 17.885l1.793 1.779-10.371 10.451L13.63 19.664l1.793-1.779 8.576 8.644z"
            />
          </svg>
        </span>
      </button>
      <div
        class="expander-list-link__main-content expander-list-link__main-content--padding"
      >
        Text 2
      </div>
    </li>
  </ul>
</div>
`;
